### PR TITLE
Rate limiting

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
Hi!

Armon had suggested that we add rate limiting to the Vault Client for the research team project. (Don't want to name it since this PR is publicly visible.) We don't want to flood Vault with requests, and he suggested we upstream this change to the main repo anyway.

Super small PR: I didn't add a test because the rate limiting functionality is provided entirely by an official Go API (which the Nomad team has been using). 

Let me know what you think. Happy to iterate on it.

### CODE REVIEW

**Comment by jefferai** [comment]
I don't really see much value in it outside of very specialized usages, but happy to talk about it offline if you want.

**Comment by RobbieMcKinstry** [comment]
Hi Jeff! Jon and I are typing up an email. Thanks! ^.^

File: api/client.go [hunk 1]
```diff
@@ -362,6 +420,15 @@ func (c *Client) Address() string {
 	return c.addr.String()
 }
 
+// SetLimiter will set the rate limiter for this client.
+// This method is thread-safe.
+// rateLimit and burst are specified according to https://godoc.org/golang.org/x/time/rate#NewLimiter
+func (c *Client) SetLimiter(rateLimit float64, burst int) {
+	c.modifyLock.RLock()
```
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply api/client.go]: Whoops! That's obvious! Sloppy from me!

[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply api/client.go]: Looks like other methods that modify the `Config` struct first lock the `Client` for reading and then lock the `Config` for writing.

I'm going to follow that pattern here.

File: api/client.go [hunk 2]
```diff
@@ -279,6 +299,38 @@ func (c *Config) ReadEnvironment() error {
 	return nil
 }
 
+func parseRateLimit(val string) (rate float64, burst int, err error) {
```
Comment (kalafut) [inline api/client.go]: Sscanf might work well here. Something like:
```
func parseRateLimit(val string) (rate float64, burst int, err error) {
	_, err = fmt.Sscanf(val, "%f:%d", &rate, &burst)
	if err != nil {
		rate, err = strconv.ParseFloat(val, 64)
		if err != nil {
			err = fmt.Errorf("%v was provided by incorrectly formatted", "EnvRateLimit")
		}
		burst = int(rate)
	}

	return rate, burst, err
}
```

Comment (RobbieMcKinstry) [reply api/client.go]: Great feedback!

Comment (RobbieMcKinstry) [reply api/client.go]: Done!

File: api/client.go [hunk 3]
```diff
@@ -362,6 +397,17 @@ func (c *Client) Address() string {
 	return c.addr.String()
 }
 
+// SetLimiter will set the rate limiter for this client.
+// This method is thread-safe.
+// rateLimit and burst are specified according to https://godoc.org/golang.org/x/time/rate#NewLimiter
+func (c *Client) SetLimiter(rateLimit float64, burst int) {
+	c.modifyLock.RLock()
+	c.config.modifyLock.Lock()
+	defer c.modifyLock.RUnlock()
```
Comment (vishalnayak) [inline api/client.go]: Can we have the order of these two defers switched?

Comment (jefferai) [reply api/client.go]: ^

File: api/client.go [hunk 4]
```diff
@@ -528,9 +575,14 @@ func (c *Client) NewRequest(method, requestPath string) *Request {
 // a Vault server not configured with this client. This is an advanced operation
 // that generally won't need to be called externally.
 func (c *Client) RawRequest(r *Request) (*Response, error) {
+	if c.config.Limiter != nil {
```
Comment (jefferai) [inline api/client.go]: This needs to happen inside the c.config RLock protected block.

File: api/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jefferai) [reply api/client.go]: Defer statements happen in reverse order, can you switch the order of these deferred calls? Or an easier to understand pattern would be:

```go
c.modifyLock.RLock()
defer c.modifyLock.RUnlock()
c.config.modifyLock.RLock()
defer c.modifyLock.RUnlock()
```

[the remark this answers is not in the document]
Comment (jefferai) [reply api/client.go]: I'm just going to merge this and fix up all the locking after.

**Comment by RobbieMcKinstry** [comment]
Thanks, everyone!

**Pull request merged**